File: practice_codes/graph.cpp
#include <iostream>
using namespace std;
struct Node 
{
    int vertex;
    Node* next;
};

struct Graph
{
    int V;
    Node** adjList; //array of pointer
    Graph(int V)
    {
        this->V = V; //no of vertices
        adjList = new Node*[V];  
        for(int i=0; i<V;i++)
            adjList[i] =nullptr; 
    }

    void addEdge(int u,int v)
    {
        Node* newNode =new Node{v,adjList[u]};
        adjList[u] = newNode;
        newNode =new Node{u,adjList[v]};
        adjList[v] =newNode;
    }

    void printGraph()
    {
        for(int i = 0; i < V; i++) {
            cout<<"Vertex "<<i<<" ->";
            Node* temp= adjList[i];
            while(temp!=nullptr){
                cout<<temp->vertex<<" ";
                temp=temp->next;
            }
            cout<<endl;
        }
    }
};

int main(){
    Graph g(5); 
    g.addEdge(0,1);
    g.addEdge(0,4);
    g.addEdge(1,2);
    g.addEdge(1,3);
    g.addEdge(1,4);
    g.addEdge(2,3);
    g.addEdge(3,4);

    g.printGraph();  
    return 0;
}

File: practice_codes/Graph.java
import java.util.*;
class Edge {   
   int src, dest;
   Edge(int src, int dest) {
      this.src = src;
      this.dest = dest;
   }
}

public class Graph {
   // node of adjacency list
   static class vertex {
      int v;
      vertex(int v) {
        this.v = v;
      }
   };

   List<List<vertex>> adj_list = new ArrayList<>();
   
   public Graph(List<Edge> edges){
      
      // adjacency list memory allocation
      for (int i = 0; i < edges.size(); i++)
      adj_list.add(i, new ArrayList<>());
      
      // add edges to the graph
      for (Edge e : edges){
        // allocate new node in adjacency List from src to dest
        adj_list.get(e.src).add(new vertex(e.dest));
      }
   }
   public static void main (String[] args) {
      List<Edge> edges = Arrays.asList(new Edge(0, 1),new Edge(0, 2),
      new Edge(0, 3),new Edge(1, 2), new Edge(1, 4),
      new Edge(2, 4), new Edge(2, 3),new Edge(3, 1));
      
      Graph graph = new Graph(edges);
      
      // print the graph as an adjacency list
      int src = 0;
      int lsize = graph.adj_list.size();
      System.out.println("The graph created is: ");
      while (src < lsize) {
      
         //traverse through the adjacency list and print the edges
         for (vertex edge : graph.adj_list.get(src)) {
            System.out.print(src + " -> " + edge.v + "\t");
         }
         System.out.println();
         src++;
      }
   }
}